### DISCUSSION TITLE
How to add metadata to stream ?

### DISCUSSION THREAD

**Discussion by etherwvlf** [savonet/liquidsoap#2243] [User Support]
This config was working fine in 1.4.4 version but after upgrading liquidsoap to last version it no longer works:

```
settings.log.stdout.set(true)
######################
def apply_metadata(m) =
  title = m["title"]
  artist = m["artist"]
  print("#{artist} - #{title}")
end
######################
playlist1 = playlist(mode="randomize",reload_mode="watch", "/dir1")
playlist2 = playlist(mode="randomize",reload_mode="watch", "/dir2")

radio = mksafe(rotate([playlist1, playlist2]))
radio = fallback([request.queue(id="request"),radio])

radio = audio_to_stereo(radio)
radio = on_metadata(apply_metadata, radio)

output.icecast(%vorbis,host = "192.168.1.5",port = 1500,password = "givememore", mount = "/mount1",radio)
```

Now it gives me:

```
user1@machine1:/home/user1$ liquidsoap lasttest.liq 
At lasttest.liq, line 16, char 8-20:
radio = on_metadata(apply_metadata, radio)

Error 4: Undefined variable on_metadata
```
I`ve tried using:
```
radio = map_metadata(apply_metadata, radio)
```
It gives:
```
user1@machine1:/home/user1$ liquidsoap lasttest.liq 
At lasttest.liq, line 16, char 21-35:
radio = map_metadata(apply_metadata, radio)

Error 5: this value has type
  (...) -> unit (inferred at lasttest.liq, line 2 char 0 - line 7 char 3)
but it should be a subtype of
  (...) -> [_]
```
How to make it work and add metadata to stream ? Sorry if the question is stupid or the solution is obvious. I am trying to make it work to no avail for the last couple of days. There are no sample configs on the Internet or in the documentation I can reference, or the solutions are outdated and no longer work.

**Comment by Vaalyn** [comment]
The `map_metadata` function requires you to return a list (indicated by the `(...) -> [_]`) so I think you'll just have to put the `m` at the bottom of the `apply_metadata` function to have it return it:

```
def apply_metadata(m) =
  title = m["title"]
  artist = m["artist"]
  print("#{artist} - #{title}")
  
  m
end

radio = map_metadata(apply_metadata, radio)
```

**Comment by etherwvlf** [comment]
Thanks a lot !
It is working now and I've never would have guessed the solution :)

**Comment by toots** [comment]
Thanks @Vaalyn for the help! If you just need to print the metadata, the easiest is to use the `on_metadata` method:
```ruby
def print_metadata(m) =
  title = m["title"]
  artist = m["artist"]
  print("#{artist} - #{title}")
end

radio.on_metadata(print_metadata)
```
`map_metadata` is provided for users who want to modify the metadata.
